src/CalculatorUITestFramework/ApplicationBase.cs

```diff
@@ -0,0 +1,205 @@
+using Microsoft.VisualStudio.TestTools.UnitTesting;
+using Microsoft.VisualStudio.TestTools.UnitTesting.Logging;
+using OpenQA.Selenium;
+using OpenQA.Selenium.Appium;
+using OpenQA.Selenium.Appium.Windows;
+using OpenQA.Selenium.Interactions;
+using OpenQA.Selenium.Remote;
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Threading;
+
+namespace CalculatorUITestFramework
+{
+    /// <summary>
+    /// The ApplicationBase class handles setting up the WinAppDriver session and gives access to UI elements that are always present on the UI regardless of calculator mode.
+    /// </summary>
+    public class ApplicationBase
+    {
+        // Note: append /wd/hub to the URL if you're directing the test at Appium
+        private const string windowsApplicationDriverUrl = "http://127.0.0.1:4723";
+        private const string calculatorAppId = "Microsoft.WindowsCalculator.Dev_8wekyb3d8bbwe!App";
+        private static Process winAppDriverProcess;
+        public static NavigationMenu NavigationMenu = new NavigationMenu();
+        public static WindowsDriver<WindowsElement> CalculatorSession { get; private set; }
+        public static WindowsElement Header
+        {
+            get
+            {
+                try
+                {
+                    return TryFindElementByAccessibilityId("Header");
+                }
+                catch
+                {
+                    return TryFindElementByAccessibilityId("ContentPresenter");
+                }
+            }
+        }
+        public static WindowsElement CalculatorResult => TryFindElementByAccessibilityId("CalculatorResults");
+
+        public ApplicationBase()
+        {
+        }
+
+        public ApplicationBase(WindowsDriver<WindowsElement> calculatorSession)
+        {
+            CalculatorSession = calculatorSession;
+        }
+
+        /// <summary>
+        /// Initalizes the WinAppDriver CalculatorSession, starts the session and launches the app.
+        /// </summary>
+        /// <param name="context">The test context passed into VSTest.</param>
+        public static void ApplicationSetup(TestContext context)
+        {
+            // Launches the WinAppDriver.exe in a new Process.
+            winAppDriverProcess = new Process();
+            winAppDriverProcess.StartInfo.FileName = @"c:\Program Files (x86)\Windows Application Driver\winappdriver.exe";
```

Comment:
Can we have a fallback of launching this ourselves?  (Or maybe and app.config that configures this).  If we don't have this here, local development gets harder.

---
In reply to: [283453418](https://github.com/microsoft/calculator/pull/501#discussion_r283453418) [](ancestors = 283453418)

Reply:
What about just hiding it behind an #if DEBUG?

Reply:
I did some more research into how other projects handle this.

If you look [upstream in appium-dotnet-driver](https://github.com/appium/appium-dotnet-driver), there's already code to start and stop Appium--see AppiumLocalService and AppiumServiceBuilder. That won't work for us since we're using WinAppDriver directly instead of full appium, but we could reuse most of that code to build WindowsDriverLocalService and WindowsDriverServiceBuilder. And eventually those patches could be contributed upstream.

As for choosing when to start the service ourselves vs. rely on the WinAppDriver Azure Pipelines task to do it for us...I don't want to make the assumption that Debug===local development in VS and Release===CI pipeline. It would be better to plumb through a flag from the build environment like $(BuildingInVisualStudio). But if we have code in our process which does all the things that AppiumLocalService does (starting and stopping the service in a robust way, reading output from stdout, etc.) then maybe there's no need for the winappdriver pipelines task any more and we don't need the conditional.

cc: @hassanuz

Reply:
I've removed the WinAppDriver process start up. I asked @joseartrivera to take a look at the AppiumLocalService and AppiumServiceBuilder.

Reply:
I added this in, as for the command line output it looks like we would only be able to see the output for failed test cases. I think this is probably fine since we would likely only dig into the logs on test failures.

See microsoft/azure-pipelines-tasks#3588